python/ray/util/sgd/v2/migration/sgd-migration-guide.md

```diff
@@ -0,0 +1,224 @@
+
+For a full list of features that Ray SGD v2 provides, please check out the user guide: https://docs.ray.io/en/master/raysgd/v2/user_guide.html. 
+
+If there are any issues or anything missing with this guide or any feedback on Ray SGD v2 overall, please file a Github issue on the Ray repo: https://github.com/ray-project/ray/issues!
+
+## What are the API differences?
+There are 2 primary API differences between Ray SGD v1 and v2.
+
+1. There is a single `Trainer` interface for all backends (torch, tensorflow, horovod), and the backend is simply specified via an argument: `Trainer(backend="torch")`, `Trainer(backend="horovod")`, etc. Any features that we add to Ray SGD will be supported for all backends, and there won't be any API divergence like there was with a separate `TorchTrainer` and `TFTrainer`.
+2. The `TrainingOperator` and creator functions are replaced by a generic user-defined training function. You no longer have to make your training logic fit into a restrictive interface. In Ray SGD v2, you simply have to provide a training function that describes the full logic for your training execution and this will be distributed by Ray SGD. 
+```python
+# Torch Example
+def train_func_distributed():
+    num_epochs = 3
+    model = NeuralNetwork()
+    model = DistributedDataParallel(model)
+    loss_fn = nn.MSELoss()
+    optimizer = optim.SGD(model.parameters(), lr=0.1)
+
+    for epoch in range(num_epochs):
+        output = model(input)
+        loss = loss_fn(output, labels)
+        optimizer.zero_grad()
+        loss.backward()
+        optimizer.step()
+        print(f"epoch: {epoch}, loss: {loss.item()}")
+
+from ray.sgd import Trainer
+
+trainer = Trainer(backend="torch", num_workers=4)
+trainer.start()
+results = trainer.run(train_func_distributed)
+trainer.shutdown()
+```
+Currently, this means that the user is now responsible for modifying their code to support distributed training (specifying `DistributedDataParallel` for `torch` or `MultiWorkerMirroredStrategy` for `tensorflow`) as opposed to having this be automatically handled internally. However, we have plans to provide utilities that you can use to automatically handle these recipes for you.
+
+## Training Logic
+The main change you will have to make is how you define your training logic. In Ray SGD v1, the API for defining training logic differed for `TorchTrainer` vs. `TFTrainer`, so the steps to migrate will be different for each of these.
+
+### PyTorch
+In v1, the training logic is defined through the `train_epoch` and `train_batch` methods of a `TrainingOperator` class which is passed into the `TorchTrainer`. To migrate to Ray SGD v2, there are 2 options: 
+1. If you liked having your training logic in the `TrainingOperator`, you can continue to use the `TrainingOperator` with Ray SGD v2.
+2. If you felt the `TrainingOperator` is too unnecessary and complex, or you had to customize it extensively, you can define your own training function.
```

Comment:
nit: Should we reverse the order of these two and encourage the user to define their own training function?  Specifically because we plan to deprecate `TrainingOperator` in the future.

Reply:
Yes good, point, made the change!